File: src/routes/cloudtrail.js
const express = require('express');
const router = express.Router();
const CloudTrailEvent = require('../models/CloudTrailEvent');

// Get all CloudTrail events with filters
router.get('/', (req, res) => {
  try {
    const filters = {
      event_name: req.query.event_name,
      workspace_id: req.query.workspace_id,
      event_from: req.query.event_from,
      event_to: req.query.event_to,
      event_source: req.query.event_source,
      limit: req.query.limit || 100,
      offset: req.query.offset || 0
    };

    const events = CloudTrailEvent.getAll(filters);

    // Parse JSON fields
    const parsedEvents = events.map(event => ({
      ...event,
      user_identity: JSON.parse(event.user_identity || '{}'),
      request_parameters: JSON.parse(event.request_parameters || '{}'),
      response_elements: JSON.parse(event.response_elements || '{}')
    }));

    res.json({
      data: parsedEvents,
      limit: parseInt(filters.limit, 10),
      offset: parseInt(filters.offset, 10)
    });
  } catch (error) {
    console.error('Error fetching events:', error);
    res.status(500).json({ error: 'Failed to fetch CloudTrail events' });
  }
});

// Get events for a specific workspace
router.get('/workspace/:workspaceId', (req, res) => {
  try {
    const events = CloudTrailEvent.getByWorkspaceId(req.params.workspaceId);
    
    const parsedEvents = events.map(event => ({
      ...event,
      user_identity: JSON.parse(event.user_identity || '{}'),
      request_parameters: JSON.parse(event.request_parameters || '{}'),
      response_elements: JSON.parse(event.response_elements || '{}')
    }));

    res.json({ data: parsedEvents });
  } catch (error) {
    console.error('Error fetching workspace events:', error);
    res.status(500).json({ error: 'Failed to fetch workspace events' });
  }
});

// Get available event names
router.get('/event-names', (req, res) => {
  try {
    const eventNames = CloudTrailEvent.getEventNames();
    res.json({ data: eventNames.map(e => e.event_name) });
  } catch (error) {
    console.error('Error fetching event names:', error);
    res.status(500).json({ error: 'Failed to fetch event names' });
  }
});

module.exports = router;

File: backend/handlers/cloudtrail.go
package handlers

import (
	"database/sql"
	"net/http"
	"strconv"

	"github.com/4syedalihassan/workspaces-inventory/models"
	"github.com/gin-gonic/gin"
)

type CloudTrailHandler struct {
	DB *sql.DB
}

// ListEvents returns CloudTrail events with filtering
func (h *CloudTrailHandler) ListEvents(c *gin.Context) {
	// Parse pagination
	limitStr := c.DefaultQuery("limit", "50")
	offsetStr := c.DefaultQuery("offset", "0")

	limit, _ := strconv.Atoi(limitStr)
	offset, _ := strconv.Atoi(offsetStr)

	// Build filters
	filters := make(map[string]interface{})

	if workspaceID := c.Query("workspace_id"); workspaceID != "" {
		filters["workspace_id"] = workspaceID
	}

	if eventName := c.Query("event_name"); eventName != "" {
		filters["event_name"] = eventName
	}

	if username := c.Query("username"); username != "" {
		filters["username"] = username
	}

	if startTime := c.Query("start_time"); startTime != "" {
		filters["start_time"] = startTime
	}

	if endTime := c.Query("end_time"); endTime != "" {
		filters["end_time"] = endTime
	}

	// Get CloudTrail events
	events, total, err := models.ListCloudTrailEvents(h.DB, filters, limit, offset)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to retrieve CloudTrail events"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"data":   events,
		"total":  total,
		"limit":  limit,
		"offset": offset,
	})
}

// GetEvent returns a single CloudTrail event by ID
func (h *CloudTrailHandler) GetEvent(c *gin.Context) {
	eventIDStr := c.Param("id")

	query := `
		SELECT id, event_id, event_name, event_time, event_source, username,
		       user_identity, workspace_id, request_parameters, response_elements,
		       event_region, created_at
		FROM cloudtrail_events
		WHERE id = $1
	`

	var event models.CloudTrailEvent
	err := h.DB.QueryRow(query, eventIDStr).Scan(
		&event.ID, &event.EventID, &event.EventName, &event.EventTime, &event.EventSource,
		&event.Username, &event.UserIdentity, &event.WorkspaceID, &event.RequestParameters,
		&event.ResponseElements, &event.EventRegion, &event.CreatedAt,
	)

	if err != nil {
		if err == sql.ErrNoRows {
			c.JSON(http.StatusNotFound, gin.H{"error": "Event not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to retrieve event"})
		return
	}

	c.JSON(http.StatusOK, event)
}

// ExportCloudTrail exports CloudTrail events to CSV or Excel
func (h *CloudTrailHandler) ExportCloudTrail(c *gin.Context) {
	format := c.DefaultQuery("format", "xlsx")

	// Build filters
	filters := make(map[string]interface{})

	if workspaceID := c.Query("workspace_id"); workspaceID != "" {
		filters["workspace_id"] = workspaceID
	}

	if eventName := c.Query("event_name"); eventName != "" {
		filters["event_name"] = eventName
	}

	if username := c.Query("username"); username != "" {
		filters["username"] = username
	}

	if startTime := c.Query("start_time"); startTime != "" {
		filters["start_time"] = startTime
	}

	if endTime := c.Query("end_time"); endTime != "" {
		filters["end_time"] = endTime
	}

	// Get all events (no pagination for export)
	events, _, err := models.ListCloudTrailEvents(h.DB, filters, 10000, 0)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to retrieve CloudTrail events"})
		return
	}

	// Export using the export handler
	ExportData(c, events, format, "cloudtrail")
}
